src/app/MessageDef.h

```diff
@@ -0,0 +1,894 @@
+};
+
+namespace AttributePath {
+enum
+{
+    kCsTag_NodeId              = 0,
+    kCsTag_EndpointId          = 1,
+    kCsTag_NamespacedClusterId = 2,
+    kCsTag_FieldId             = 3,
+    kCsTag_ListIndex           = 4,
+};
+
+class Parser : public ParserBase
+{
+public:
+    // aReader has to be on the element of AttributePath container
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
+
+    // Roughly verify the schema is right, including
+    // 1) all mandatory tags are present
+    // 2) no unknown tags
+    // 3) all elements have expected data type
+    // 4) any tag can only appear once
+    CHIP_ERROR CheckSchemaValidity() const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetNodeId(chip::NodeId * const apNodeId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetEndpointId(chip::EndpointId * const apEndpointId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetNamespacedClusterId(chip::ClusterId * const apClusterId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetFieldId(uint8_t * const apFieldId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetListIndex(uint16_t * const apListIndex) const;
+};
+
+class Builder : public BuilderBase
+{
+public:
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter, const uint8_t aContextTagToUse);
+
+    AttributePath::Builder & NodeId(const chip::NodeId aNodeId);
+    AttributePath::Builder & EndpointId(const chip::EndpointId aEndpointId);
+    AttributePath::Builder & NamespacedClusterId(const chip::ClusterId aNamespacedClusterId);
+    AttributePath::Builder & FieldId(const uint8_t aFieldId);
+    AttributePath::Builder & ListIndex(const uint16_t aListIndex);
+
+    AttributePath::Builder & EndOfAttributePath();
+
+private:
+    CHIP_ERROR _Init(chip::TLV::TLVWriter * const apWriter, const uint64_t aTag);
+};
+
+}; // namespace AttributePath
+
+namespace AttributePathList {
+class Parser : public ListParserBase
+{
+public:
+    // Roughly verify the schema is right, including
+    // 1) at least one element is there
+    // 2) all elements are anonymous and of Path type
+    // 3) every path is also valid in schema
+    CHIP_ERROR CheckSchemaValidity() const;
```

Comment:
can this similarly be on the base class?

Reply:
CheckSchemaValidity implementation is different for different subclass, for example, commandpath, attributepath

Reply:
every class has to have an implementation, but every class doesn't need to declare the method...

Reply:
If we don't declare CheckSchemaValidity in derived class, AttributePath::Parser::ParserBase::CheckSchemaValidity() ==> AttributePath::Parser::Parser::CheckSchemaValidity()  seems not work. In addition, I have to add explicit scope for tags such as kCsTag_NodeId in CheckSchemaValidity assume we are in AttributePath  since the function is not in derived class.

I would create one ticket to track this issue.

Reply:
https://github.com/project-chip/connectedhomeip/issues/3944